### PR TITLE
Key rotation logic for jobs in NuGet.Jobs

### PR DESCRIPTION
**Pull request by scottbommarito**
Replaced the ordinary `Dictionary<string, string>` that stored config arguments with a `RefreshingArgumentsDictionary` which refreshes at an interval that can be passed in through the command line.

Additionally, job initialization is now done with every run of the job loop, so that each job gets the updated config values when they are refreshed.

I have tested that this works with Search.GenerateAuxilliaryData and looked at the rest of the jobs to make sure there is no caching of the `RefreshingArgumentsDictionary` by jobs (which would lead to key rotation bugs), but given the sheer number of jobs affected by this change I was unsure how to methodically test them.

### CODE REVIEW

**Comment by maartenba** [comment]
LGTM!

Now in case of key rotation, new values are only picked up after a day, right? (worst-case)

File: src/NuGet.Jobs.Common/RefreshingArgumentsDictionary.cs [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace NuGet.Jobs
+{
+    public class RefreshingArgumentsDictionary : IDictionary<string, string>
+    {
+        private const int DefaultRefreshIntervalSec = 86400;
```
Comment (shishirx34) [inline src/NuGet.Jobs.Common/RefreshingArgumentsDictionary.cs]: add comment `// 24 hours`

File: src/NuGet.Jobs.Common/RefreshingArgumentsDictionary.cs [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+                        return _injectedArguments[key].Item1;
+                    }
+                }
+                
+                return processArgument(key).Result;
+            }
+
+            private set
+            {
+                _injectedArguments[key] = new Tuple<string, DateTime>(value, DateTime.Now);
```
Comment (shishirx34) [inline src/NuGet.Jobs.Common/RefreshingArgumentsDictionary.cs]: Lets use `.UtcNow` to standardize everywhere.

File: src/NuGet.Jobs.Common/RefreshingArgumentsDictionary.cs [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+        public string this[string key]
+        {
+            get
+            {
+                if (!_unprocessedArguments.ContainsKey(key)) throw new KeyNotFoundException();
+
+                if (_injectedArguments.ContainsKey(key))
+                {
+                    if (DateTime.Now.Subtract(_injectedArguments[key].Item2).TotalSeconds < _refreshArgsIntervalSec)
```
Comment (shishirx34) [inline src/NuGet.Jobs.Common/RefreshingArgumentsDictionary.cs]: `.UtcNow`

File: src/NuGet.Jobs.Common/RefreshingArgumentsDictionary.cs [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+            catch(KeyNotFoundException)
+            {
+                value = "";
+                return false;
+            }
+        }
+
+        private async Task<string> processArgument(string key)
+        {
+            var value = await _secretInjector.InjectAsync(_unprocessedArguments[key]);
```
Comment (shishirx34) [inline src/NuGet.Jobs.Common/RefreshingArgumentsDictionary.cs]: nit: `argumentValue` instead of `value`

File: src/NuGet.Jobs.Common/RefreshingArgumentsDictionary.cs [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+            set
+            {
+                throw new NotImplementedException();
+            }
+        }
+
+        public void Add(string key, string value)
+        {
+            throw new NotImplementedException();
```
Comment (shishirx34) [inline src/NuGet.Jobs.Common/RefreshingArgumentsDictionary.cs]: what if there is a need to add a new <key, value> pair, I think we should implement this.

**Comment by shishirx34** [comment]
:watch:

**Comment by shishirx34** [comment]
Question: Are you going to use the same Changes to `JobConfigurationManager` in other jobs too? for Nuget.Services.Metadata?

**Comment by scottbommarito** [comment]
@shishirx34 I will use the same changes if they do things in the same way. As far as I know, NuGet.Staging uses a different method, but if the rest use JobConfigurationManager then I'll probably make a very similar change there.

File: src/NuGet.Jobs.Common/RefreshingArgumentsDictionary.cs (hunk 5 above)
Comment (scottbommarito) [inline src/NuGet.Jobs.Common/RefreshingArgumentsDictionary.cs]: I don't think we'll ever need to use it but since you already wrote this code I don't mind throwing it in there.

Comment (shishirx34) [inline src/NuGet.Jobs.Common/RefreshingArgumentsDictionary.cs]: Also we should have unit tests for the `RefreshingDictionary`

**Comment by shishirx34** [comment]
The reason I asked is because, there we use some caching mechanism, which won't work the same way we do here. I guess we can cross that bridge when we get there.

**Comment by scottbommarito** [comment]
@maartenba Yes, if we happened to change the keys just after the jobs tried to refresh, it would not refresh until after the refresh interval had elapsed. The default for that right now is a day, but if we decide we want to shorten or lengthen that, I also added support for specifying your own interval through the command line argument `-RefreshArgsIntervalSec`.

**Comment by scottbommarito** [comment]
@shishirx34 If there's already a cache, it'll probably make my job easier lol

File: src/NuGet.Jobs.Common/RefreshingArgumentsDictionary.cs [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+        }
+
+        public void CopyTo(KeyValuePair<string, string>[] array, int arrayIndex)
+        {
+            throw new NotImplementedException();
+        }
+
+        public bool Remove(KeyValuePair<string, string> item)
+        {
+            throw new NotImplementedException();
```
Comment (shishirx34) [inline src/NuGet.Jobs.Common/RefreshingArgumentsDictionary.cs]:
```
return Remove(item.key);
```

**Comment by shishirx34** [comment]
One more comment, Looks Good otherwise!

**Comment by ryuyu** [comment]
:shipit:

**Comment by scottbommarito** [comment]
@skofman1 @maartenba @ryuyu @xavierdecoster @shishirx34 

I rewrote a large amount of the logic to fix possible async issues as well as moved the refreshing arguments dictionary to the KeyVault project in preparation of its use elsewhere. Unfortunately this required changes to each Job.Init(). Mind taking another look?

File: src/Gallery.CredentialExpiration/Job.cs [hunk 7]
```diff
@@ -39,40 +40,40 @@ public class Job : JobBase
...
-                var instrumentationKey = JobConfigurationManager.TryGetArgument(jobArgsDictionary, JobArgumentNames.InstrumentationKey);
+                var instrumentationKey = await jobArgsDictionary.GetOrDefault<string>(JobArgumentNames.InstrumentationKey);
                 ApplicationInsights.Initialize(instrumentationKey);
 
                 var loggerConfiguration = LoggingSetup.CreateDefaultLoggerConfiguration(ConsoleLogOnly);
                 var loggerFactory = LoggingSetup.CreateLoggerFactory(loggerConfiguration);
                 _logger = loggerFactory.CreateLogger<Job>();
 
-                _whatIf = JobConfigurationManager.TryGetBoolArgument(jobArgsDictionary, JobArgumentNames.WhatIf);
+                _whatIf = await jobArgsDictionary.GetOrDefault<bool>(JobArgumentNames.WhatIf);
```
Comment (ryuyu) [inline src/Gallery.CredentialExpiration/Job.cs]: >GetOrDefault<bool> [](start = 50, length = 18)

Maybe have this take a parameter to specify a default value to return?

File: src/NuGet.Jobs.Common/JobRunner.cs [hunk 8]
```diff
@@ -60,24 +61,29 @@ static JobRunner()
...
 
                 // Set JobTraceListener. This will be done on every job run as well
-                SetJobTraceListener(job, consoleLogOnly, jobArgsDictionary);
+                await SetJobTraceListener(job, consoleLogOnly, jobArgsDictionary);
 
-                bool runContinuously = !JobConfigurationManager.TryGetBoolArgument(jobArgsDictionary, JobArgumentNames.Once);
-                int? sleepDuration = JobConfigurationManager.TryGetIntArgument(jobArgsDictionary, JobArgumentNames.Sleep); // sleep is in milliseconds
-                if (!sleepDuration.HasValue)
+                bool runContinuously = !await jobArgsDictionary.GetOrDefault<bool>(JobArgumentNames.Once);
+                int sleepDuration = await jobArgsDictionary.GetOrDefault<int>(JobArgumentNames.Sleep); // sleep is in milliseconds
```
Comment (shishirx34) [inline src/NuGet.Jobs.Common/JobRunner.cs]: I think this should still be nullable? what if you want to specify 0 as a value? I think the default value should always be `null` unless specified the default like @ryuyu mentioned.

File: src/NuGet.Jobs.Common/JobRunner.cs [hunk 9]
```diff
@@ -60,24 +61,29 @@ static JobRunner()
...
+                if (sleepDuration != default(int))
```
Comment (shishirx34) [inline src/NuGet.Jobs.Common/JobRunner.cs]: I think this is in counter to it was previously. It used to check if `sleepDuration` didn't have a value. Which in your case should translate to `sleepDuration == default(int)`. Another case in point why returning a `null` is useful for GetOrDefault<int>. I think in all cases unless specified by another argument, the default should return `null`. Same way TryGetArguement used to do.

File: src/NuGet.Jobs.Common/JobRunner.cs [hunk 10]
```diff
@@ -60,24 +61,29 @@ static JobRunner()
...
                 {
-                    sleepDuration = JobConfigurationManager.TryGetIntArgument(jobArgsDictionary, JobArgumentNames.Interval);
-                    if (sleepDuration.HasValue)
+                    sleepDuration = await jobArgsDictionary.GetOrDefault<int>(JobArgumentNames.Interval);
+                    if (sleepDuration != default(int))
```
Comment (shishirx34) [inline src/NuGet.Jobs.Common/JobRunner.cs]: here you have it correctly as opposed to the one above.

File: src/NuGet.Jobs.Common/JobRunner.cs [hunk 11]
```diff
@@ -60,24 +61,29 @@ static JobRunner()
...
                     {
-                        sleepDuration = sleepDuration.Value * 1000; // interval is in seconds
+                        sleepDuration = sleepDuration * 1000; // interval is in seconds
+                    }
+                    else
+                    {
+                        Trace.TraceWarning("SleepDuration is not provided or is not a valid integer. Unit is milliSeconds. Assuming default of 5000 ms...");
+                        sleepDuration = 5000;
```
Comment (shishirx34) [inline src/NuGet.Jobs.Common/JobRunner.cs]: why do you need to set this value? What used to happen in previous cases? I thought it did use some default value.

If you do need to keep this, set the value as a static const field of `5000`

File: src/NuGet.Jobs.Common/JobRunner.cs (hunk 8 above)
Comment (scottbommarito) [inline src/NuGet.Jobs.Common/JobRunner.cs]: Generics and `Nullable` do not play well together. I can specify the default as `-1` and then check for that.

Comment (scottbommarito) [inline src/NuGet.Jobs.Common/JobRunner.cs]: (also, I would preferred to have maintained the `TryGet...` pattern, but unfortunately `async` doesn't allow `ref` or `out` parameters so it's impossible)

File: src/NuGet.Jobs.Common/JobRunner.cs (hunk 9 above)
Comment (scottbommarito) [inline src/NuGet.Jobs.Common/JobRunner.cs]: you're correct, simple mistake, should be `==`

File: src/NuGet.Jobs.Common/JobRunner.cs (hunk 11 above)
Comment (scottbommarito) [inline src/NuGet.Jobs.Common/JobRunner.cs]: added a constant

File: src/NuGet.Jobs.Common/JobRunner.cs (hunk 8 above)
Comment (shishirx34) [inline src/NuGet.Jobs.Common/JobRunner.cs]: You misunderstood what I was trying to say. I meant similar to `TryGet...` simply return `null` for all cases as the default value. Leave the assertion to the code and not worry about what the actual data type's default value is. All the `TryGet...`s apart from `TryGetBoolArgument` returns null as the default. So you can keep to that standard.

Comment (scottbommarito) [inline src/NuGet.Jobs.Common/JobRunner.cs]: Talked offline. Due to limitations in the language, it's impossible to get the `TryGet` or `Nullable` patterns working with both generics and `async`. It would be possible to write wrapper functions for each type, essentially getting rid of the generics, but not sure if we want to do that.

Paging Dr @skofman1 for her opinion on the matter.

File: src/Gallery.CredentialExpiration/Job.cs (hunk 7 above)
Comment (scottbommarito) [inline src/Gallery.CredentialExpiration/Job.cs]: This has been done, but this particular instance does not have a default specified because it is redundant (`default(bool)` is `false`).

**Comment by xavierdecoster** [comment]
@scottbommarito Once this one is merged, can you please open the same PR against the new ServerCommon repo? https://github.com/NuGet/ServerCommon/tree/develop Want to make sure your changes also end up in the new packages as NuGet.Services.KeyVault is going to be removed from this repo :)

**Comment by scottbommarito** [comment]
@xavierdecoster I'll probably merge the ServerCommon one and then update the references in this to point to that before I merge this!

File: src/NuGet.Services.KeyVault/SecretInjector.cs [hunk 12]
```diff
@@ -36,6 +36,8 @@ public SecretInjector(ISecretReader secretReader, string frame)
 
         public async Task<string> InjectAsync(string input)
         {
+            if (input == null) return input;
```
Comment (skofman1) [inline src/NuGet.Services.KeyVault/SecretInjector.cs]: >input == null [](start = 16, length = 13)

string.IsNullOrEmpty

File: src/NuGet.Services.KeyVault/RefreshingArgumentsDictionary.cs [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+        {
+            _secretInjector = secretInjector;
+            _unprocessedArguments = unprocessedArguments;
+            _injectedArguments = new Dictionary<string, Tuple<string, DateTime>>();
+
+            var refreshArgsIntervalSec = DefaultRefreshIntervalSec;
+            if (_unprocessedArguments.ContainsKey(RefreshArgsIntervalSec))
+            {
+                int parsedRefreshInterval;
+                if (int.TryParse(_unprocessedArguments[RefreshArgsIntervalSec], out parsedRefreshInterval)) refreshArgsIntervalSec = parsedRefreshInterval;
```
Comment (skofman1) [inline src/NuGet.Services.KeyVault/RefreshingArgumentsDictionary.cs]: >if (int.TryParse(_unprocessedArguments[RefreshArgsIntervalSec], out parsedRefreshInterval)) [](start = 16, length = 91)

use {}. No one line ifs. It's not readable.

File: src/NuGet.Services.KeyVault/RefreshingArgumentsDictionary.cs [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+        /// Gets an argument from the cache.
+        /// If the argument has not been cached or needs to be refreshed, it reinjects the KeyVault secrets and updates the cache.
+        /// </summary>
+        /// <param name="key">The key associated with the desired argument.</param>
+        /// <returns>The argument associated with the given key.</returns>
+        /// <exception cref="KeyNotFoundException">Thrown when the key is not found in the list of arguments.</exception>
+        /// <exception cref="ArgumentNullException">Thrown when the argument associated with the given key is null or empty.</exception>
+        private async Task<string> Get(string key)
+        {
+            if (!_unprocessedArguments.ContainsKey(key)) throw new KeyNotFoundException();
```
Comment (skofman1) [inline src/NuGet.Services.KeyVault/RefreshingArgumentsDictionary.cs]: >new [](start = 63, length = 3)

same.

File: src/NuGet.Services.KeyVault/RefreshingArgumentsDictionary.cs [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+
+            if (!_injectedArguments.ContainsKey(key) || DateTime.UtcNow.Subtract(_injectedArguments[key].Item2).TotalSeconds >= _refreshArgsIntervalSec)
+            {
+                await processArgument(key);
+            }
+
+            var argumentValue = _injectedArguments[key].Item1;
+            if (string.IsNullOrEmpty(argumentValue)) throw new ArgumentNullException();
```
Comment (skofman1) [inline src/NuGet.Services.KeyVault/RefreshingArgumentsDictionary.cs]: >ArgumentNullException [](start = 63, length = 21)

same

File: src/NuGet.Services.KeyVault/RefreshingArgumentsDictionary.cs [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+        public const string RefreshArgsIntervalSec = "CacheRefreshInterval";
+        private const int DefaultRefreshIntervalSec = 60 * 60 * 24; // 1 day (24 hrs)
+
+        private int _refreshArgsIntervalSec;
+        private ISecretInjector _secretInjector;
+
+        private Dictionary<string, string> _unprocessedArguments;
+        private Dictionary<string, Tuple<string, DateTime>> _injectedArguments;
+
+        public RefreshingArgumentsDictionary(ISecretInjector secretInjector, Dictionary<string, string> unprocessedArguments)
```
Comment (skofman1) [inline src/NuGet.Services.KeyVault/RefreshingArgumentsDictionary.cs]: >Dictionary<string, string> unprocessedArguments [](start = 77, length = 47)

Since this code is general, and not used only in Jobs, you shouldn't assume Dictionary as configuration provider. 
Options to configure a class:
1. Pass RefreshArgsIntervalSec in ctor
2. Expose RefreshArgsIntervalSec as settable property if u want it to be changed at runtime
3. Create RefreshingArgumentsDictionaryConfiguration class that will contain all config value u want, and pass in ctor.

File: src/NuGet.Services.KeyVault/RefreshingArgumentsDictionary.cs (hunk 14 above)
Comment (skofman1) [inline src/NuGet.Services.KeyVault/RefreshingArgumentsDictionary.cs]: >KeyNotFoundException [](start = 67, length = 20)

add details to exception, otherwise we have no clue which key we were looking for

File: src/NuGet.Services.KeyVault/RefreshingArgumentsDictionary.cs [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+        {
+            _injectedArguments[key] = new Tuple<string, DateTime>(value, DateTime.UtcNow);
+        }
+
+        /// <summary>
+        /// Fetches the argument from KeyVault and updates the cache.
+        /// </summary>
+        /// <param name="key">The key associated with the desired argument.</param>
+        /// <returns>The argument, freshly updated from KeyVault.</returns>
+        private async Task<string> processArgument(string key)
```
Comment (skofman1) [inline src/NuGet.Services.KeyVault/RefreshingArgumentsDictionary.cs]: >p [](start = 35, length = 1)

P
(not java ;) )

File: src/NuGet.Services.KeyVault/RefreshingArgumentsDictionary.cs [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+                await processArgument(key);
+            }
+
+            var argumentValue = _injectedArguments[key].Item1;
+            if (string.IsNullOrEmpty(argumentValue)) throw new ArgumentNullException();
+
+            return argumentValue;
+        }
+
+        public async Task<T> GetOrThrow<T>(string key)
```
Comment (skofman1) [inline src/NuGet.Services.KeyVault/RefreshingArgumentsDictionary.cs]: >GetOrThrow [](start = 29, length = 10)

why is this needed here? Who uses it?

File: src/NuGet.Services.KeyVault/RefreshingArgumentsDictionary.cs [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+            catch (KeyNotFoundException)
+            {
+                // The specified key was not found in the arguments.
+            }
+            catch (NotSupportedException)
+            {
+                // Could not convert an object of type string into an object of type T.
+            }
+            return defaultValue;
+        }
```
Comment (skofman1) [inline src/NuGet.Services.KeyVault/RefreshingArgumentsDictionary.cs]: again, this is nice, but unrelated to the basic logic of this code. Why is it needed here? Can we move it out of this class?

File: src/NuGet.Services.KeyVault/RefreshingArgumentsDictionary.cs [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+
+        public void Set(string key, string value)
+        {
+            _injectedArguments[key] = new Tuple<string, DateTime>(value, DateTime.UtcNow);
```
Comment (skofman1) [inline src/NuGet.Services.KeyVault/RefreshingArgumentsDictionary.cs]: >_injectedArguments [](start = 12, length = 18)

u add values here to _injectedArguments but ContainsKey checks against _unprocessedArguments?

What is the usage scenario? Do we need it?

File: src/NuGet.Services.KeyVault/RefreshingArgumentsDictionary.cs (hunk 18 above)
Comment (scottbommarito) [inline src/NuGet.Services.KeyVault/RefreshingArgumentsDictionary.cs]: This is the "this configuration is required and we should fail if we can't find it" method. It's used very frequently.

File: src/NuGet.Services.KeyVault/RefreshingArgumentsDictionary.cs (hunk 20 above)
Comment (scottbommarito) [inline src/NuGet.Services.KeyVault/RefreshingArgumentsDictionary.cs]: This is mostly for testing--the Get logic should handle setting and refreshing the arguments correctly. I use this method in certain tests where we need to guarantee a specific config.

This probably should set both injected and unprocessed, however, so I'll fix that.

File: src/NuGet.Services.KeyVault/RefreshingArgumentsDictionary.cs (hunk 17 above)
Comment (scottbommarito) [inline src/NuGet.Services.KeyVault/RefreshingArgumentsDictionary.cs]: I was wondering why this method looked weird!

File: src/NuGet.Services.KeyVault/RefreshingArgumentsDictionary.cs (hunk 19 above)
Comment (scottbommarito) [inline src/NuGet.Services.KeyVault/RefreshingArgumentsDictionary.cs]: It's used pretty much everywhere that this class is used. Basically, `GetOrThrow` is the method for required arguments and `GetOrDefault` is the method for optional arguments. It wouldn't make much sense to separate the logic out because almost every time we use configuration we have both required and optional parameters.

File: src/NuGet.Services.KeyVault/RefreshingArgumentsDictionary.cs (hunk 20 above)
Comment (shishirx34) [inline src/NuGet.Services.KeyVault/RefreshingArgumentsDictionary.cs]: I think the `Set` here should set the value only on _unprocessedArguments, and the next time the Get is called on this key the Get method will take care of setting the values in _injectedArguments.

Comment (scottbommarito) [inline src/NuGet.Services.KeyVault/RefreshingArgumentsDictionary.cs]: I thought of that, but I think if we are `Set`'ing something that has an `_injectedArguments` then it would still fetch the previously cached value, not the newly `Set` value.

**Comment by scottbommarito** [comment]
A cleaner solution for this was found, since it's not necessary to make everything async because the deadlock can only occur in MVC apps.

https://github.com/NuGet/NuGet.Jobs/pull/68
